SwissDataScienceCenter/renku-ui
# feat(server): websocket server

This PR adds backend support for WebSocket.
The client can now create a new channel and send commands. The server will verify the user identity and send relevant information to the client based on the commands received.

Mind that this first implementation: 
* Supports only a short and a long loop, repeating approximately every 5 seconds and 3 minutes.
* Cache the data locally since sharing it through Redis might lead to out-of-sync actions (that's easy to change if we wish).
* Tries to optimize performances by providing a single response to every channel (multiple channels would be created by opening multiple tabs in the browser).

**How to test**
The WebSocket channel communication can be started by reaching the endpoint `wss://<URL>/ui-server/ws`. Postman can be used for that.

You might want to set up authentication (not strictly required) by adding the header `Cookie` with value `ui-server-session=<your-valid-cookie>` (you can find the valid cookie in your browser after logging in and checking the stored cookies).

After connecting, you should get a message like the one at the end of the following screenshot

![Screenshot_20220822_172346](https://user-images.githubusercontent.com/43481553/185958617-8b7bb444-511f-454c-bb3c-b7f2a0b6a0c6.png)

You can now send commands in the format of `{   "timestamp", "type": "init", "data": {}   }`.
Try with the following one:

```
{
    "timestamp": 1,
    "type": "init",
    "data": {
        "requestServerVersion": "true"
    }
}
```

If you are running telepresence, you will get a negative response (see the following screenshot). Otherwise, you will get the current UI version and be notified every 3 minutes on the same channel.

![Screenshot_20220822_172555](https://user-images.githubusercontent.com/43481553/185959095-1316f0b4-df7d-405e-8d23-2baf73314a36.png)


/deploy #persist
re #1338
